Cause copy module to fail on empty string as source

samdoran opened the pull request:
##### SUMMARY

Cause `copy` to fail rather than copying the content of the current working directory if an empty string is specified as the source. Fixes #27363

Cleaned up formatting on integration tests.

Added integration test for this specific case.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request


##### COMPONENT NAME

`action/copy.py`

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.4
```


##### ADDITIONAL INFORMATION

This has been a bug for quite some time. I don't believe fixing this will have any adverse effects as it seems it was the original intent to fail on no input as the source.

samdoran committed:
Fail if an empty string is set as src for copy module

Fixes #27363

samdoran committed:
Cleanup task formatting on copy tests

Use multi-line YAML
Add debug statements with verbosity: 1 rather than leave them in there commented out.

samdoran committed:
Add test for empty string as source

dagwieers commented on lib/ansible/plugins/action/copy.py:
```diff
@@ -407,7 +407,7 @@ def run(self, tmp=None, task_vars=None):
 
         result['failed'] = True
-        if (source is None and content is None) or dest is None:
+        if ((not source or source is None) and content is None) or dest is None:
```
I would split it out and have more specific error messages:
```python
    if not source and content is None:
        result['msg'] = "src (or content) is required"
    elif dest is not None:
        result['msg'] = "dest is required"
    elif source and content is not None:
        result['msg'] = "src and content are mutually exclusive"
```
Can dest be an empty string ? (current script directory ?)

dagwieers replied:
PS I also modified the last conditional in that list.

samdoran committed:
Do more checks in order to add more specific errors messages

Add more integration tests for the various failure scenarios.
Cleanup some syntax on existing integration test tasks.

dagwieers commented on lib/ansible/plugins/action/copy.py:
```diff
@@ -407,12 +407,14 @@ def run(self, tmp=None, task_vars=None):
-            result['msg'] = "src (or content) and dest are required"
-        elif source is not None and content is not None:
-            result['msg'] = "src and content are mutually exclusive"
+        if (not source or source is None) and content is None:
```
If source is None, the conditional `not source` will be true. So I don't think the second part of this conditional will ever hit.

samdoran replied:
Good point. This was what I was originally trying to fix with this PR. I just tested with only `not source` and it passed the tests. Will push shortly.

Please try to merge this today.

dagwieers reviewed (approved):
Good stuff. Check my one note, and then we can merge this.
